standalone/node/src/chain_spec.rs

```diff
@@ -61,9 +61,50 @@ pub struct Extensions {
 
 /// Specialized `ChainSpec`.
 pub type ChainSpec = sc_service::GenericChainSpec<
-	GenesisConfig,
+	GenesisExt,
 	Extensions,
 >;
+
+/// Extension for the Phala devnet genesis config to support a custom changes to the genesis state.
+#[derive(serde::Serialize, serde::Deserialize)]
+pub struct GenesisExt {
+	/// The runtime genesis config.
+	runtime_genesis_config: GenesisConfig,
+	/// The block duration in milliseconds.
+	///
+	/// If `None` is supplied, the default value is used.
+	block_milliseconds: Option<u64>,
+}
+
+impl sp_runtime::BuildStorage for GenesisExt {
+	fn assimilate_storage(&self, storage: &mut sp_core::storage::Storage) -> Result<(), String> {
+		sp_state_machine::BasicExternalities::execute_with_storage(storage, || {
+			if let Some(bm) = self.block_milliseconds.as_ref() {
+				MillisecsPerBlock::set(bm);
+				let bm_f = *bm as f64;
```

Comment:
Do you think we can replace all the floats by integers?

I understand it may not introduce any non-deterministic in this PR, because everything is pre-calculated before the genesis block, but it still looks a bit weird here... If integer will introduce more magic, I'm happy with the current implementation, but if not, I still prefer pure integers.

Reply:
The floating point logic was introduced to make sure that `secs_per_block` (and subsequent `minutes`, `hours` and `days`) calculation is performed properly.

If we just used `bm: u64` instead of `bm_f: f64` in the calculation for `secs_per_block`, we would lose precision for any `bm % 1000 != 0`.

For example, if `bm == 1500`, we would have:
- `secs_per_block == 1`
- `minutes == 60`
- `hours == 3600`
- `days == 86400`

On the other hand, the floating point logic makes sure that if `bm == 1500`, we have:
- `secs_per_block == 1.5`
- `minutes == 40`
- `hours == 2400`
- `days == 57600`

Therefore, we conclude that avoiding floating points introduces inaccuracies into the runtime.

Reply:
Acknowledged